modules/pipeline/providers/dialeragent/interface.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dialeragent
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/rancher/remotedialer"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/pkg/discover"
+)
+
+type Interface interface {
+	RegisterEdgeToDialer(ctx context.Context)
+}
+
+// RegisterEdgeToDialer only registers the edge to the dialer
+func (p *provider) RegisterEdgeToDialer(ctx context.Context) {
+	if !p.Cfg.IsEdge {
+		p.Log.Warnf("current node is not edge node, skip dialer agent registration")
+		return
+	}
+	if p.Cfg.ClusterAccessKey == "" {
+		p.Log.Errorf("cluster access key is empty, couldn't make edge registration")
+		return
+	}
+	edgeDetail := apistructs.ClusterDialerClientDetail{
+		apistructs.ClusterDialerDataKeyPipelineAddr: p.Cfg.PipelineAddr,
+		apistructs.ClusterDialerDataKeyPipelineHost: p.Cfg.PipelineHost,
+	}
+	edgeDetailBytes, _ := edgeDetail.Marshal()
+	headers := http.Header{
+		apistructs.ClusterDialerHeaderKeyClusterKey.String():    {p.Cfg.ClusterName},
+		apistructs.ClusterDialerHeaderKeyClientType.String():    {apistructs.ClusterDialerClientTypePipeline.String()},
+		apistructs.ClusterDialerHeaderKeyAuthorization.String(): {p.Cfg.ClusterAccessKey},
+		apistructs.ClusterDialerHeaderKeyClientDetail.String():  {string(edgeDetailBytes)},
+	}
+	ep, err := p.parseDialerEndpoint()
+	if err != nil {
+		p.Log.Errorf("failed to parse dialer endpoint: %v", err)
+		return
+	}
+	for {
+		// if client connect successfully, will block until client disconnect
+		err = remotedialer.ClientConnect(ctx, ep, headers, nil, p.ConnectAuthorizer, nil)
+		if err != nil {
+			p.Log.Errorf("failed to connect to dialer: %v", err)
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(p.Cfg.RetryConnectDialerInterval):
+			// retry connect dialer
```

Comment:
maybe retry directly is better.
use select.default.

Reply:
use select.default maybe too frequently， set default interval 1 second